README
NAME
    Metabase - A database framework and API for resource metadata

VERSION
    version 1.002

DESCRIPTION
    Metabase is a database framework and API for resource metadata. The
    framework describes how arbitrary data ("facts") are associated with
    particular resources or related to each other. The API describes how to
    store, retrieve, and search this information.

  History and Motivation
    Metabase was originally designed as a means of storing reports from the
    CPAN Testers project. When Metabase was initially developed, CPAN
    Testers reports were sent by individual testers to a single email
    server, which then forwarded them to a USENET group, which was
    considered the authoritative store. This presented problems: some
    testers couldn't send email, the system wasn't very searchable, was hard
    to mirror, and the data inside the system was entirely unstructured.

    Metabase aimed to avoid all of those problems by being
    transport-neutral, searchable and easier to mirror by design, and geared
    toward storing structured data. Simplicity is another design goal: while
    it has several moving parts, they're all simple and designed to be
    replaceable and extensible, rather than to be a perfect design up front.

OVERVIEW
    A Metabase has several parts:

    *   Metabase::Librarian, a class which manages access to the Archive and
        Index

    *   Metabase::Gateway, a role for managing submission of facts to the
        Librarian

    *   Metabase::Archive, a role for storing and retrieving facts

    *   Metabase::Index, a role for indexing and searching facts

    *   Metabase::Query, a role for translating a common query data
        structure into a backend-specific query

    Roles require implementations. These could use flat files, relational
    databases, object databases, cloud services, or anything else that can
    satisfy the role API.

    Metabase comes with some simple, stupid backends for testing:

    *   Metabase::Archive::Filesystem

    *   Metabase::Index::FlatFile

    Facts stored with in a Metabase are defined as subclasses of
    Metabase::Fact. Metabase::Report is a subclass that relates multiple
    facts.

    Metabase::Web provides the web API for storing, searching and retrieving
    facts. Metabase::Client::Simple is the client library to submit facts to
    a Metabase::Web server. A future Metabase::Client class will provide
    submit and search capabilities.

SEE ALSO
    *   Metabase::Backend::AWS

    *   Metabase::Backend::MongoDB

    *   Metabase::Backend::SQL

SUPPORT
  Bugs / Feature Requests
    Please report any bugs or feature requests through the issue tracker at
    <https://rt.cpan.org/Public/Dist/Display.html?Name=Metabase>. You will
    be notified automatically of any progress on your issue.

  Source Code
    This is open source software. The code repository is available for
    public review and contribution under the terms of the license.

    <https://github.com/dagolden/metabase>

      git clone git://github.com/dagolden/metabase.git

AUTHORS
    *   David Golden <dagolden@cpan.org>

    *   Ricardo Signes <rjbs@cpan.org>

    *   Leon Brocard <acme@cpan.org>

CONTRIBUTORS
    *   Florian Ragwitz <rafl@debian.org>

    *   Graham Knop <haarg@haarg.org>

    *   Leon Brocard <acme@astray.com>

    *   Ricardo SIGNES <rjbs@cpan.org>

COPYRIGHT AND LICENSE
    This software is Copyright (c) 2013 by David Golden.

    This is free software, licensed under:

      The Apache License, Version 2.0, January 2004



Changes
Revision history for Metabase

1.002     2013-07-16 10:36:43 America/New_York
    - Fix a test that had been broken by 5.18's hash ordering changes

1.001     2012-05-18 15:34:47 America/New_York

    [FIXED]

    - Fixed bug where default CHI wasn't initialized properly

1.000     2012-05-16 13:23:04 America/New_York

    - The 1.000 release reflects a major reorganization of the Metabase
      distribution and is not backwards compatible.

    - The Archive and Index roles have an expanded API with additional
      required methods

    - A Query role has been added to abstract query requirements from the
      indexer to allow for backend-specific query implementations

    - Tests have been converted to Test::Routine style roles, which are
      released with the distribution for use in backend tests in other
      distributions.

    - AWS and SQLite backends have been removed and will be released as
      separate distributions.

    - Solr backend has been removed.

0.016     2010-12-07 15:27:49 EST5EDT

    - No changes.  Released as stable

0.015     2010-09-28 09:41:27 EST5EDT

    - Add explicit dependency on SQL::Translator 0.11006 for DBIx::Class
      deploy() used in Metabase::Archive::SQLite

0.014     2010-09-20 22:22:45 EST5EDT

    - Now uses SimpleDB::Client instead of SimpleDB::Class

    - Updated dist.ini to @DAGOLDEN style

    - Prerequisites now auto-detected

0.013     2010-09-01 18:29:24 EST5EDT

    - Improve diagnostics on failure

    - Fix occasional failures in t/librarian/report.t due to time
      comparison between original and saved reports

    - Add iterator generators to librarian (only supported on S3 so far)

    - Load fact classes on demand

    - Require latest Metabase::Fact (0.018)

0.012     2010-04-24 15:18:33 EST5EDT

    - Fix authentication cache bug (first check was always failing)

0.011     2010-04-23 23:38:58 EST5EDT

    - Fix cache attribute type constraint

    - Cache authentication strings, not objects

0.010     2010-04-23 17:40:39 EST5EDT

    - Add authentication caching.  Adds CHI as a prereq

    - Properly document Metabase::Gateway as a Moose::Role

0.009     2010-04-22 07:26:46 EST5EDT

    - Don't add undefined values to the index

0.008     2010-04-20 09:54:07 EST5EDT

    - Fact content metadata keys no longer have to be lowercase

0.007     2010-04-11 12:00:49 EST5EDT

    - Fix a bug in Metabase::Archive::S3 -- NOTE: THIS WILL BREAK
      ACCESS TO OLD S3 FACTS

0.006     2010-03-29 14:29:25 EST5EDT

    - JSON will now always encode/decode to/from ASCII

    - Converted to Dist::Zilla for easier release management

0.005

    - Remove dependency on JSON::XS and rely on JSON to use it
      if installed (see Task::Metabase)

    - Ensure JSON encoding includes utf8 encoding to avoid
      errors compressing wide characters in Metabase::Archive::S3

0.004 Mon Mar 15 22:30:34 EDT 2010
    - Fixed -limit operator for SimpleDB search
    - Reduced numerous prerequisites to lower verisons

0.003 Mon Mar  8 23:42:52 EST 2010
    - heavily revised for CPAN Testers 2.0 effort
    - Metabase::Gateway is now a Moose::Role
    - Added Amazon Web Service backends

0.002 (never released)
    - private development version, never released to CPAN

0.001 24 Jun 2009
    - original skeleton created at Oslo QA hackathon 2008


Makefile.PL

use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "A database framework and API for resource metadata",
  "AUTHOR" => "David Golden <dagolden\@cpan.org>, Ricardo Signes <rjbs\@cpan.org>, Leon Brocard <acme\@cpan.org>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Metabase",
  "EXE_FILES" => [],
  "LICENSE" => "apache",
  "NAME" => "Metabase",
  "PREREQ_PM" => {
    "CHI" => 0,
    "CPAN::DistnameInfo" => 0,
    "Carp" => 0,
    "Class::MOP" => 0,
    "Data::GUID" => 0,
    "Data::Stream::Bulk::Array" => 0,
    "Data::Stream::Bulk::Callback" => 0,
    "Data::Stream::Bulk::Filter" => 0,
    "Data::Stream::Bulk::Nil" => 0,
    "Fcntl" => 0,
    "File::Slurp" => 0,
    "IO::File" => 0,
    "JSON" => 2,
    "List::AllUtils" => 0,
    "Metabase::Fact" => "0.018",
    "Metabase::Fact::String" => 0,
    "Metabase::User::Profile" => 0,
    "Metabase::User::Secret" => 0,
    "Moose" => "1.00",
    "Moose::Role" => 0,
    "Moose::Util::TypeConstraints" => 0,
    "MooseX::Types::Moose" => 0,
    "MooseX::Types::Path::Class" => 0,
    "MooseX::Types::Structured" => 0,
    "Regexp::SQL::LIKE" => "0.001",
    "Test::Deep" => 0,
    "Test::More" => "0.92",
    "Test::Routine" => 0,
    "Tie::File" => 0,
    "namespace::autoclean" => 0,
    "parent" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Find" => 0,
    "File::Path" => 0,
    "File::Spec::Functions" => 0,
    "File::Temp" => 0,
    "List::Util" => 0,
    "Metabase::Report" => 0,
    "Path::Class" => 0,
    "Test::Exception" => 0,
    "Test::Routine::Util" => 0,
    "base" => 0,
    "lib" => 0
  },
  "VERSION" => "1.002",
  "test" => {
    "TESTS" => "t/*.t t/archive/*.t t/gateway/*.t t/index/*.t t/librarian/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



